apis/projectcontour/v1/httpproxy.go

```diff
@@ -377,6 +377,23 @@ type Route struct {
 	// RequestRedirectPolicy defines an HTTP redirection.
 	// +optional
 	RequestRedirectPolicy *HTTPRequestRedirectPolicy `json:"requestRedirectPolicy,omitempty"`
+
+	// DirectResponsePolicy returns an arbitrary HTTP response directly.
+	// +optional
+	DirectResponsePolicy *HTTPDirectResponsePolicy `json:"directResponsePolicy,omitempty"`
+}
+
+type HTTPDirectResponsePolicy struct {
+	// StatusCode is the HTTP response status to be returned.
+	// +required
+	// +kubebuilder:validation:Minimum=200
+	// +kubebuilder:validation:Maximum=599
+	StatusCode int `json:"statusCode"`
+
+	// Body is the content of the response body.
+	// If this setting is omitted, no body is included in the generated response.
+	// +optional
```

Comment:
should we set a kubebuilder validation max length on this? something somewhat arbitrary for now to be defensive?

Reply:
> should we set a kubebuilder validation max length on this? something somewhat arbitrary for now to be defensive?

@sunjayBhatia It is a good idea to limit the maximum length of the the direct response body, do you have any good suggestions for this?

Reply:
> should we set a kubebuilder validation max length on this? something somewhat arbitrary for now to be defensive?

What do you think about setting the max length to 1048576？

Reply:
> hm you may hit the default etcd/k8s max object size of 1mb before getting to that 1mb limit on a string field, which does limit how large of a body you can serve anyways
> 
> since contour will have to load that resource into memory and also send that over to envoy, maybe it would be better to discourage such a large body to start and in fact choose something quite a bit smaller, since it can significantly impact resource usage, especially in a multitenant installation of contour

@sunjayBhatia  Yes, 1mb is sometimes a waste of resources for configuration distribution. But if this value is set too small, it may not meet the requirements.
This requires us to strike a balance between the use of resources and the user experience.

Reply:
I think a high-ish limit plus a note in the field description makes sense -- it's worth specifically calling out that the contents will have to be sent over xDS *every time* there is any config change, since we're doing SOTW, so could lead to increased latency & resource utilization. I do see your point @sunjayBhatia about a multi-tenant install, but on balance I guess I'd lean towards letting folks have a large response if they need it, provided we've documented what they should watch out for.

Maybe the best of all worlds would be to have a configurable size limit, so Contour admins could limit the max size to something smaller if needed. Probably could be a follow-on enhancement though, if we wanted to do this.

Reply:
Adding a note to this field is a great way, I will add relevant note in this field. It can be discussed later whether it needs to be limited by configuration or gatekeeper.

Reply:
ping @sunjayBhatia @skriss

Reply:
added https://github.com/projectcontour/contour/issues/4547 for a follow up